### ISSUE TITLE
AKSampler crash in MultiSegmentEnvelopeGenerator::skipEmptySegments()

### ISSUE THREAD

**Issue by quantamrhino** [AudioKit/AudioKit#2224]
I'm on v5 master latest.

I simply load a AVAudiofile in AKSampler and then play it.
I don't change the ADSR values.
This worked fine in 4.10.

```
bool MultiSegmentEnvelopeGenerator::skipEmptySegments() //skips over any segment w/ length 0, so as to not influence the state of the envelope
    {
        SegmentDescriptor seg = (*segments)[curSegIndex]; <<<---- crash here. segments is null, curSegIndex is 2
        int length = seg.lengthSamples;
        while (length == 0) { //skip any segments that are 0-length
            curSegIndex++;
            seg = (*segments)[curSegIndex];
            length = seg.lengthSamples;
        }
        if (curSegIndex >= int(segments->size())) // if at end of the envelope, reset
        {
            reset(segments);
            return false;
        }
        return true;
    }
```

**Comment by aure** [comment]
v5 is not officially but thanks for checking it out. AKSampler shouldn't have changed too much between 4.10 and 5.0beta, but if you want to track down the difference, that would be great.

**Comment by wtholliday** [comment]
@quantamrhino can you write a little unit test to reproduce this?

**Comment by quantamrhino** [comment]
I have a repro (attached zip file - drag files to a empty xcode project). I'm using v5 beta1. (Podfile also attached)

I see at least 4 issues:

1. If sample rate is not set, the sample (which s 48K) plays
2. but sounds nothing like the original file.
2. It also never stops. Even after calling stop() on the sampler. (there's a play and stop button in the UI)
3. If sample rate is set to 48K, the sampler crashes in the way said in bug above.

[sampler bug.zip](https://github.com/AudioKit/AudioKit/files/5069037/sampler.bug.zip)

**Comment by aure** [comment]
I have just changed AKSampler to disallow loading audio files or sfz except during initialization. This helps resolve the problem with the segments.  I realize that disallowing loading files is probably bad for certain apps that rely on this feature in changing presets for instance (I believe our own apps use this). But, its not very good practice. Instead, I think a better idea would be to use a mixer, plug the sampler into it, and when you want to change, use mixer.removeInput(oldSampler) and mixer.addInput(newSampler). Wouldn't you agree @wtholliday?

As for the sample rate, AudioKit has 44.1K hardcoded in a number of places and this may also be a problem here. I have tested your code rewritten for AK5 v5-develop and at 44.1K and it seems to work as advertised.  Unless @getdunne can chime in or until more info is provided I will close this issue.
